File: Why Is My App Not Responding_/apply-migrations.mjs
import mysql from 'mysql2/promise';

const url = process.env.DATABASE_URL;
if (!url) { console.error('DATABASE_URL not set'); process.exit(1); }

const conn = await mysql.createConnection(url);

const statements = [
  // flashcards
  `CREATE TABLE IF NOT EXISTS \`flashcards\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`studyMaterialId\` int,
    \`courseId\` int,
    \`front\` text NOT NULL,
    \`back\` text NOT NULL,
    \`easeFactor\` float NOT NULL DEFAULT 2.5,
    \`interval\` int NOT NULL DEFAULT 0,
    \`repetitions\` int NOT NULL DEFAULT 0,
    \`nextReviewAt\` timestamp NOT NULL DEFAULT (now()),
    \`lastReviewedAt\` timestamp,
    \`correctCount\` int NOT NULL DEFAULT 0,
    \`incorrectCount\` int NOT NULL DEFAULT 0,
    \`difficulty\` enum('easy','medium','hard') NOT NULL DEFAULT 'medium',
    \`tags\` json,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    \`updatedAt\` timestamp NOT NULL DEFAULT (now()) ON UPDATE CURRENT_TIMESTAMP,
    CONSTRAINT \`flashcards_id\` PRIMARY KEY(\`id\`)
  )`,
  // study_materials
  `CREATE TABLE IF NOT EXISTS \`study_materials\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`courseId\` int,
    \`title\` varchar(500) NOT NULL,
    \`type\` enum('audio','pdf','document_scan','video') NOT NULL,
    \`originalFileUrl\` text,
    \`originalFileKey\` text,
    \`processedFileUrl\` text,
    \`processedFileKey\` text,
    \`status\` enum('pending','processing','completed','failed') NOT NULL DEFAULT 'pending',
    \`processingError\` text,
    \`duration\` int,
    \`pageCount\` int,
    \`wordCount\` int,
    \`language\` varchar(10),
    \`transcription\` text,
    \`summary\` text,
    \`keyPoints\` json,
    \`definitions\` json,
    \`timestamps\` json,
    \`subject\` varchar(100),
    \`course\` varchar(200),
    \`tags\` json,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    \`updatedAt\` timestamp NOT NULL DEFAULT (now()) ON UPDATE CURRENT_TIMESTAMP,
    CONSTRAINT \`study_materials_id\` PRIMARY KEY(\`id\`)
  )`,
  // study_sessions
  `CREATE TABLE IF NOT EXISTS \`study_sessions\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`sessionType\` enum('review','flashcards','reading','practice') NOT NULL,
    \`studyMaterialId\` int,
    \`startedAt\` timestamp NOT NULL DEFAULT (now()),
    \`endedAt\` timestamp,
    \`durationMinutes\` int,
    \`itemsReviewed\` int NOT NULL DEFAULT 0,
    \`itemsCorrect\` int NOT NULL DEFAULT 0,
    \`itemsIncorrect\` int NOT NULL DEFAULT 0,
    \`subject\` varchar(100),
    \`notes\` text,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    CONSTRAINT \`study_sessions_id\` PRIMARY KEY(\`id\`)
  )`,
  // subscription_history
  `CREATE TABLE IF NOT EXISTS \`subscription_history\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`action\` enum('subscribed','upgraded','downgraded','canceled','renewed') NOT NULL,
    \`fromTier\` enum('free','student','scholar','academic'),
    \`toTier\` enum('free','student','scholar','academic') NOT NULL,
    \`amount\` float,
    \`currency\` varchar(3) DEFAULT 'usd',
    \`stripeEventId\` varchar(255),
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    CONSTRAINT \`subscription_history_id\` PRIMARY KEY(\`id\`)
  )`,
  // weakness_analyses
  `CREATE TABLE IF NOT EXISTS \`weakness_analyses\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`testImageUrl\` text NOT NULL,
    \`testImageKey\` text NOT NULL,
    \`testType\` varchar(100),
    \`subject\` varchar(100),
    \`mistakes\` json,
    \`weakConcepts\` json,
    \`predictions\` json,
    \`score\` float,
    \`totalQuestions\` int,
    \`correctAnswers\` int,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    \`updatedAt\` timestamp NOT NULL DEFAULT (now()) ON UPDATE CURRENT_TIMESTAMP,
    CONSTRAINT \`weakness_analyses_id\` PRIMARY KEY(\`id\`)
  )`,
  // courses
  `CREATE TABLE IF NOT EXISTS \`courses\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`name\` varchar(200) NOT NULL,
    \`code\` varchar(50),
    \`color\` varchar(7) NOT NULL DEFAULT '#6366f1',
    \`semester\` varchar(50),
    \`instructor\` varchar(200),
    \`description\` text,
    \`materialsCount\` int NOT NULL DEFAULT 0,
    \`flashcardsCount\` int NOT NULL DEFAULT 0,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    \`updatedAt\` timestamp NOT NULL DEFAULT (now()) ON UPDATE CURRENT_TIMESTAMP,
    CONSTRAINT \`courses_id\` PRIMARY KEY(\`id\`)
  )`,
  // user_preferences
  `CREATE TABLE IF NOT EXISTS \`user_preferences\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`userId\` int NOT NULL,
    \`themeMode\` enum('light','dark','auto') DEFAULT 'dark',
    \`purpleShade\` varchar(20) DEFAULT 'default',
    \`accentColor\` varchar(7) DEFAULT '#a855f7',
    \`fontSize\` enum('small','medium','large') DEFAULT 'medium',
    \`sidebarPosition\` enum('left','right','hidden') DEFAULT 'left',
    \`cardLayout\` enum('compact','comfortable','spacious') DEFAULT 'comfortable',
    \`dashboardLayout\` varchar(50) DEFAULT 'default',
    \`experimentalFeatures\` text,
    \`submittedSuggestions\` text,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    \`updatedAt\` timestamp NOT NULL DEFAULT (now()) ON UPDATE CURRENT_TIMESTAMP,
    CONSTRAINT \`user_preferences_id\` PRIMARY KEY(\`id\`),
    CONSTRAINT \`user_preferences_userId_unique\` UNIQUE(\`userId\`)
  )`,
  // material_questions
  `CREATE TABLE IF NOT EXISTS \`material_questions\` (
    \`id\` int AUTO_INCREMENT NOT NULL,
    \`materialId\` int NOT NULL,
    \`userId\` int NOT NULL,
    \`question\` text NOT NULL,
    \`answer\` text NOT NULL,
    \`tokensUsed\` int,
    \`responseTime\` int,
    \`createdAt\` timestamp NOT NULL DEFAULT (now()),
    CONSTRAINT \`material_questions_id\` PRIMARY KEY(\`id\`)
  )`,
];

// ALTER users table - add columns if not exist
const userAlters = [
  { col: 'subscriptionTier', sql: "ALTER TABLE `users` ADD COLUMN `subscriptionTier` enum('free','student','scholar','academic','founder','beta') NOT NULL DEFAULT 'free'" },
  { col: 'subscriptionStatus', sql: "ALTER TABLE `users` ADD COLUMN `subscriptionStatus` enum('active','canceled','past_due','trialing') DEFAULT 'active'" },
  { col: 'stripeCustomerId', sql: "ALTER TABLE `users` ADD COLUMN `stripeCustomerId` varchar(255)" },
  { col: 'stripeSubscriptionId', sql: "ALTER TABLE `users` ADD COLUMN `stripeSubscriptionId` varchar(255)" },
  { col: 'subscriptionEndsAt', sql: "ALTER TABLE `users` ADD COLUMN `subscriptionEndsAt` timestamp" },
  { col: 'monthlyAudioMinutes', sql: "ALTER TABLE `users` ADD COLUMN `monthlyAudioMinutes` int DEFAULT 0 NOT NULL" },
  { col: 'monthlyPdfPages', sql: "ALTER TABLE `users` ADD COLUMN `monthlyPdfPages` int DEFAULT 0 NOT NULL" },
  { col: 'usageResetAt', sql: "ALTER TABLE `users` ADD COLUMN `usageResetAt` timestamp DEFAULT (now()) NOT NULL" },
  { col: 'userTier', sql: "ALTER TABLE `users` ADD COLUMN `userTier` enum('regular','founder_core','beta_tester') DEFAULT 'regular' NOT NULL" },
  { col: 'seatNumber', sql: "ALTER TABLE `users` ADD COLUMN `seatNumber` int" },
  { col: 'seatType', sql: "ALTER TABLE `users` ADD COLUMN `seatType` enum('reserved_family','public_founder','beta_tester')" },
  { col: 'lifetimeFree', sql: "ALTER TABLE `users` ADD COLUMN `lifetimeFree` boolean DEFAULT false NOT NULL" },
  { col: 'betaFreeYearUnlocked', sql: "ALTER TABLE `users` ADD COLUMN `betaFreeYearUnlocked` boolean DEFAULT false NOT NULL" },
  { col: 'betaFreeYearStart', sql: "ALTER TABLE `users` ADD COLUMN `betaFreeYearStart` timestamp" },
  { col: 'betaFreeYearEnd', sql: "ALTER TABLE `users` ADD COLUMN `betaFreeYearEnd` timestamp" },
  { col: 'founderBadge', sql: "ALTER TABLE `users` ADD COLUMN `founderBadge` varchar(100)" },
];

console.log('Creating tables...');
for (const sql of statements) {
  try {
    await conn.execute(sql);
    console.log('  OK:', sql.substring(0, 60) + '...');
  } catch (e) {
    console.log('  SKIP (already exists?):', e.message.substring(0, 80));
  }
}

console.log('\nAltering users table...');
// First check existing columns
const [cols] = await conn.execute("SHOW COLUMNS FROM `users`");
const existingCols = new Set(cols.map(c => c.Field));

for (const { col, sql } of userAlters) {
  if (existingCols.has(col)) {
    console.log(`  SKIP: ${col} already exists`);
    continue;
  }
  try {
    await conn.execute(sql);
    console.log(`  OK: Added ${col}`);
  } catch (e) {
    console.log(`  ERR: ${col}: ${e.message.substring(0, 80)}`);
  }
}

// Verify
const [tables] = await conn.execute("SHOW TABLES");
console.log('\nFinal tables:', tables.map(t => Object.values(t)[0]).join(', '));

await conn.end();
console.log('Done!');

File: Why Is My App Not Responding_/apply_migrations.py
#!/usr/bin/env python3
"""
Generate a Node.js migration script from Drizzle SQL migration files.

Usage:
  python apply_migrations.py <project_path>

Reads all .sql files in <project_path>/drizzle/, extracts CREATE TABLE IF NOT EXISTS
and ALTER TABLE statements, and writes an apply-migrations.mjs script to <project_path>.

Run the generated script with: cd <project_path> && node apply-migrations.mjs
"""

import sys
import os
import re
import glob

def main():
    if len(sys.argv) < 2:
        print("Usage: python apply_migrations.py <project_path>")
        sys.exit(1)

    project_path = sys.argv[1]
    drizzle_dir = os.path.join(project_path, "drizzle")

    if not os.path.isdir(drizzle_dir):
        print(f"Error: {drizzle_dir} not found")
        sys.exit(1)

    sql_files = sorted(glob.glob(os.path.join(drizzle_dir, "*.sql")))
    if not sql_files:
        print("No .sql migration files found in drizzle/")
        sys.exit(0)

    # Collect all SQL from migration files
    all_sql = []
    for f in sql_files:
        with open(f) as fh:
            content = fh.read()
            # Replace drizzle statement breakpoints with semicolons
            content = content.replace("--> statement-breakpoint", ";")
            all_sql.append(content)

    combined = "\n".join(all_sql)

    # Extract CREATE TABLE statements
    creates = re.findall(
        r'(CREATE TABLE\s+(?:IF NOT EXISTS\s+)?`\w+`\s*\([^;]+\))',
        combined, re.DOTALL | re.IGNORECASE
    )

    # Extract ALTER TABLE statements
    alters = re.findall(
        r'(ALTER TABLE\s+`\w+`\s+(?:ADD|MODIFY|DROP|CHANGE)[^;]+)',
        combined, re.DOTALL | re.IGNORECASE
    )

    print(f"Found {len(creates)} CREATE TABLE statements")
    print(f"Found {len(alters)} ALTER TABLE statements")

    # Write the migration runner
    output_path = os.path.join(project_path, "apply-migrations.mjs")
    with open(output_path, "w") as out:
        out.write('import mysql from "mysql2/promise";\n\n')
        out.write('const url = process.env.DATABASE_URL;\n')
        out.write('if (!url) { console.error("DATABASE_URL not set"); process.exit(1); }\n\n')
        out.write('const conn = await mysql.createConnection(url);\n\n')

        # Write CREATE TABLE statements
        out.write("const creates = [\n")
        for stmt in creates:
            # Ensure IF NOT EXISTS
            if "IF NOT EXISTS" not in stmt.upper():
                stmt = stmt.replace("CREATE TABLE", "CREATE TABLE IF NOT EXISTS", 1)
            escaped = stmt.replace("\\", "\\\\").replace("`", "\\`").replace("${", "\\${")
            out.write(f"  `{escaped}`,\n")
        out.write("];\n\n")

        # Write ALTER TABLE statements
        out.write("const alters = [\n")
        for stmt in alters:
            escaped = stmt.replace("\\", "\\\\").replace("`", "\\`").replace("${", "\\${")
            out.write(f"  `{escaped}`,\n")
        out.write("];\n\n")

        out.write('console.log("Creating tables...");\n')
        out.write("for (const sql of creates) {\n")
        out.write("  try {\n")
        out.write("    await conn.execute(sql);\n")
        out.write('    console.log("  OK:", sql.substring(0, 60) + "...");\n')
        out.write("  } catch (e) {\n")
        out.write('    console.log("  SKIP:", e.message.substring(0, 80));\n')
        out.write("  }\n")
        out.write("}\n\n")

        out.write('console.log("\\nApplying ALTER statements...");\n')
        out.write("for (const sql of alters) {\n")
        out.write("  try {\n")
        out.write("    await conn.execute(sql);\n")
        out.write('    console.log("  OK:", sql.substring(0, 60) + "...");\n')
        out.write("  } catch (e) {\n")
        out.write('    console.log("  SKIP:", e.message.substring(0, 80));\n')
        out.write("  }\n")
        out.write("}\n\n")

        out.write('const [tables] = await conn.execute("SHOW TABLES");\n')
        out.write('console.log("\\nFinal tables:", tables.map(t => Object.values(t)[0]).join(", "));\n')
        out.write("await conn.end();\n")
        out.write('console.log("Done!");\n')

    print(f"Generated: {output_path}")
    print(f"Run with: cd {project_path} && node apply-migrations.mjs")

if __name__ == "__main__":
    main()
